kse/src/org/kse/gui/dialogs/DGeneratingDHParameters.java

```diff
@@ -0,0 +1,213 @@
+		jbCancel.getActionMap().put(CANCEL_KEY, new AbstractAction() {
+			private static final long serialVersionUID = 1L;
+
+			@Override
+			public void actionPerformed(ActionEvent evt) {
+				cancelPressed();
+			}
+		});
+
+		Container pane = getContentPane();
+		pane.setLayout(new MigLayout("insets dialog", "[]","[]unrel"));
+		pane.add(jlGenDHParameters, "wrap");
+		pane.add(jpbGenDHParameters, "growx, wrap");
+		pane.add(jbCancel, "tag Cancel");
+
+		addWindowListener(new WindowAdapter() {
+			@Override
+			public void windowClosing(WindowEvent evt) {
+				if ((generator != null) && (generator.isAlive())) {
+					generator.interrupt();
+				}
+				closeDialog();
+			}
+		});
+
+		setTitle(res.getString("DGeneratingDHParameters.Title"));
+		setResizable(false);
+
+		pack();
+	}
+
+	/**
+	 * Start DH Parameters generation in a separate thread.
+	 */
+	public void startDHParametersGeneration() {
+		generator = new Thread(new GenerateDHParameters());
+		generator.setPriority(Thread.MIN_PRIORITY);
+		generator.start();
+	}
+	
+	/**
+	 * Returns the current success status
+	 * 
+	 * @return successStatus The success status boolean
+	 */
+	public boolean isSuccessful() {
+		return successStatus;
+	}
+    
+	/**
+	 * Calls the close dialogue,
+	 * Sets the success value to false
+	 */
+	private void cancelPressed() {
+		if ((generator != null) && (generator.isAlive())) {
+			generator.interrupt();
+		}
+		successStatus = false;
+		closeDialog();
+	}
+    
+	/**
+	 * Closes the dialogue
+	 */
+	private void closeDialog() {
+		setVisible(false);
+		dispose();
+	}
+
+	/**
+	 * Get the generated DH Parameters.
+	 *
+	 * @return byte array of the generated DH Parameters or null
+	 * if the user cancelled the dialog or an error occurred
+	 */
+	public byte[] getDHParameters() {
+		return dhParameters;
+	}
+
+	/**
+	 * Generates the DH Parameters
+	 * 
+	 * Identifies a safe prime using the Bouncy Castle provider. 
+	 *
+	 */
+	private class GenerateDHParameters implements Runnable {
+		@Override
+		public void run() {
+			try {
+				AlgorithmParameterGenerator	algGen = AlgorithmParameterGenerator.getInstance("DH",BOUNCY_CASTLE.jce());
+	    		algGen.init(keySize, new SecureRandom());
+	    		AlgorithmParameters dhParams = algGen.generateParameters();	    		
+	    		DHParameterSpec dhSpec = (DHParameterSpec)dhParams.getParameterSpec(DHParameterSpec.class);   		
+
+	    		// Generator G is set as random in params, but it has to be 2 to conform to openssl
+	    	    DHParameters realParams = new DHParameters(dhSpec.getP(), BigInteger.valueOf(2));	    	    
+	    	    // Add DH Params to ASN.1 Encoding vector
+	    	    ASN1EncodableVector vec = new ASN1EncodableVector();
+	    	    vec.add(new ASN1Integer(realParams.getP()));
+	    	    vec.add(new ASN1Integer(realParams.getG()));
+	    	    // Add DER Encoding to byte array
+	    	    byte [] derEncodedBytes = new DERSequence(vec).getEncoded(ASN1Encoding.DER);
+                dhParameters = derEncodedBytes;
```

Comment:
There is a mix of tabs and spaces here

Reply:
Oh, and lines 192 and 193 can be combined into one line, the variable `derEncodedBytes` is not necessary

Reply:
Updated.